src/inc/til/env.h

```diff
@@ -0,0 +1,515 @@
+                                {
+                                    data = {
+                                        reinterpret_cast<wchar_t*>(valueData.data()), valueData.size() / sizeof(wchar_t)
+                                    };
+                                }
+                                else if (pass == 1 && (type == REG_EXPAND_SZ) && valueDataSize >= sizeof(wchar_t))
+                                {
+                                    data = {
+                                        reinterpret_cast<wchar_t*>(valueData.data()), valueData.size() / sizeof(wchar_t)
+                                    };
+                                    data = expand_environment_strings(data.data());
+                                }
+
+                                // Because Registry data may or may not be null terminated... check if we've managed
+                                // to store an extra null in the wstring by telling it to create itself from pointer and size.
+                                // If we did, pull it off.
+                                if (!data.empty())
+                                {
+                                    if (data.back() == L'\0')
+                                    {
+                                        data = data.substr(0, data.size() - 1);
+                                    }
+                                }
+
+                                if (!data.empty())
+                                {
+                                    if (is_path_var(valueName))
+                                    {
+                                        concat_var(valueName, std::wstring{ data });
+                                    }
+                                    else
+                                    {
+                                        set_user_environment_var(valueName, std::wstring{ data });
+                                    }
+                                }
+                            }
+
+                            valueName.resize(maxValueNameSize);
+                            valueData.resize(maxValueDataSize);
+                            valueNameSize = maxValueNameSize;
+                            valueDataSize = maxValueDataSize;
+                            index++;
+                        }
+                        index = 0;
+                    }
+                }
+            }
+        }
+
+        std::wstring expand_environment_strings(std::wstring input)
+        {
+            // TODO: this should be replacing from ourselves, not from the OS
+            return wil::ExpandEnvironmentStringsW<std::wstring, 256>(input.data());
+        }
+
+        void set_user_environment_var(std::wstring var, std::wstring value)
+        {
+            value = expand_environment_strings(value);
+            value = check_for_temp(value);
+            save_to_map(var, value);

```

Comment:
when we call save_to_map, we should consider always `std::move`ing the inputs _as long as they're not intended to live forever_. That saves us on copying the string for every step of the call.

Then inside save_to_map, we can `insert_or_assign(std::move(var), ...)` to save yet another copy :D

Reply:
(you don't need to `move` the ones you construct on-the-spot using `wstring{}`)